src/tree/updater_gpu_hist.cu

```diff
@@ -282,63 +440,109 @@ struct DeviceShard {
         null_gidx_value(n_bins),
         param(param),
         prediction_cache_initialised(false) {
-    // Convert to ELLPACK matrix representation
+    // copy cuts to the GPU
+    dh::safe_cuda(cudaSetDevice(device_idx));
+    thrust::device_vector<float> cuts_d(hmat.cut);
+    thrust::device_vector<size_t> cut_row_ptrs_d(hmat.row_ptr);
+
+    // find the maximum row size
+    thrust::device_vector<size_t> row_ptr_d
+      (row_batch.ind_ptr + row_begin, row_batch.ind_ptr + row_end + 1);
+
     int max_elements_row = 0;
-    for (auto i = row_begin; i < row_end; i++) {
-      max_elements_row =
-          (std::max)(max_elements_row,
-                     static_cast<int>(gmat.row_ptr[i + 1] - gmat.row_ptr[i]));
-    }
+    int block = 256;
+    thrust::device_vector<int> max_row_size_d(1, 0);
+    max_row_size_k<<<dh::DivRoundUp(n_rows, block), block>>>
+      (max_row_size_d.data().get(), row_ptr_d.data().get(), n_rows);
```

Comment:
Could this kernel be replaced with something like this:
```
int n = 10;
thrust::device_vector<size_t> offsets(n + 1);
thrust::sequence(offsets.begin(), offsets.end(), 0, 5);
auto segments = thrust::make_zip_iterator(
    thrust::make_tuple(offsets.begin(), offsets.begin() + 1));
auto get_size = [=] __device__(thrust::tuple<size_t, size_t> segment) {
  return thrust::get<1>(segment) - thrust::get<0>(segment);
};
using TransformT = thrust::transform_iterator<decltype(get_size),
                                              decltype(segments), size_t>;
TransformT row_size = TransformT(segments, get_size);
size_t max =
    thrust::reduce(row_size, row_size + n, 0, thrust::maximum<size_t>());
fprintf(stderr, "max %d\n", max);
```
If you used a cub reduce this would give state-of-the-art performance and be much safer. This function could be slightly modified to accept a binary op so you could do a max reduce: https://github.com/dmlc/xgboost/blob/49b9f39818d9002c3fffcd00da0bec6d573f4f17/src/common/device_helpers.cuh#L810

Reply:
Done (using thrust).